sopel/plugins/capabilities.py

```diff
@@ -0,0 +1,455 @@
+
+        Each element is a capability request as a tuple of capability names::
+
+            >>> manager.acknowledged
+            {('cap1',), ('cap2', 'cap3')}
+
+        An acknowledged request is a registered and requested request for which
+        the bot received a ``CAP ACK`` message.
+
+        Only requested requests can be acknowledged.
+        """
+        return frozenset(self._acknowledged)
+
+    @property
+    def denied(self) -> FrozenSet[Tuple[str, ...]]:
+        """Set of denied capability requests.
+
+        Each element is a capability request as a tuple of capability names::
+
+            >>> manager.denied
+            {('cap1',), ('cap2', 'cap3')}
+
+        A denied request is a registered and requested request for which the
+        bot received a ``CAP NAK`` message.
+
+        Only requested requests can be denied.
+        """
+        return frozenset(self._denied)
+
+    @property
+    def is_complete(self) -> bool:
+        """Tell if the capability negotiation is complete.
+
+        When capability negotiation is complete, the bot can send ``CAP END``
+        to notify the server that negotiation is complete.
+
+        The capability negotiation is complete when all capability requests
+        have been either acknowledged or denied successfuly (directly or by
+        calling the :meth:`resume` method).
+        """
+        return all(
+            status
+            for cap_req, values in self._registered.items()
+            for _plugin_capability, status in values.values()
+            if cap_req in self._requested
+        )
+
+    # tell if registered, requested, acknowledged, or denied
+
+    def is_registered(self, request: Iterable[str]) -> bool:
+        """Tell if a capability request is registered.
+
+        :param request: a set of capabilities that form a capability request
+                        together; this can be any sortable iterable
+        """
+        return tuple(sorted(request)) in self._registered
+
+    def is_requested(self, request: Iterable[str]) -> bool:
+        """Tell if a capability request is requested.
+
+        :param request: a set of capabilities that form a capability request
+                        together; this can be any sortable iterable
+        """
+        return tuple(sorted(request)) in self._requested
+
+    def is_acknowledged(self, request: Iterable[str]) -> bool:
+        """Tell if a capability request is acknowledged.
+
+        :param request: a set of capabilities that form a capability request
+                        together; this can be any sortable iterable
+        """
+        return tuple(sorted(request)) in self._acknowledged
+
+    def is_denied(self, request: Iterable[str]) -> bool:
+        """Tell if a capability request is denied.
+
+        :param request: a set of capabilities that form a capability request
+                        together; this can be any sortable iterable
```

Comment:
All (finite) iterables are compatible with `sorted()`

```suggestion
    def is_registered(self, request: Iterable[str]) -> bool:
        """Tell if a capability request is registered.

        :param request: a set of capabilities that form a capability request
                        together; this can be any iterable
        """
        return tuple(sorted(request)) in self._registered

    def is_requested(self, request: Iterable[str]) -> bool:
        """Tell if a capability request is requested.

        :param request: a set of capabilities that form a capability request
                        together; this can be any iterable
        """
        return tuple(sorted(request)) in self._requested

    def is_acknowledged(self, request: Iterable[str]) -> bool:
        """Tell if a capability request is acknowledged.

        :param request: a set of capabilities that form a capability request
                        together; this can be any iterable
        """
        return tuple(sorted(request)) in self._acknowledged

    def is_denied(self, request: Iterable[str]) -> bool:
        """Tell if a capability request is denied.

        :param request: a set of capabilities that form a capability request
                        together; this can be any iterable
```